src/app/icd/client/CheckInHandler.cpp

```diff
@@ -0,0 +1,191 @@
+{}
+
+CHIP_ERROR CheckInHandler::Init(Messaging::ExchangeManager * exchangeManager, ICDClientStorage * clientStorage,
+                                CheckInDelegate * delegate)
+{
+    VerifyOrReturnError(exchangeManager != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(clientStorage != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(mpExchangeManager == nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mpICDClientStorage == nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    mpExchangeManager  = exchangeManager;
+    mpICDClientStorage = clientStorage;
+    mpCheckInDelegate  = delegate;
+
+    return mpExchangeManager->RegisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::ICD_CheckIn, this);
+}
+
+void CheckInHandler::Shutdown()
+{
+    mpICDClientStorage = nullptr;
+    mpCheckInDelegate  = nullptr;
+    if (mpExchangeManager)
+    {
+        mpExchangeManager->UnregisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::ICD_CheckIn);
+        mpExchangeManager = nullptr;
+    }
+}
+
+CHIP_ERROR CheckInHandler::OnUnsolicitedMessageReceived(const PayloadHeader & payloadHeader, ExchangeDelegate *& newDelegate)
+{
+    // Return error for wrong message type
+    VerifyOrReturnError(payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::ICD_CheckIn),
+                        CHIP_ERROR_INVALID_MESSAGE_TYPE);
+
+    newDelegate = this;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CheckInHandler::OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                             System::PacketBufferHandle && payload)
+{
+    // If the message type is not ICD_CheckIn, return CHIP_NO_ERROR and exit
+    VerifyOrReturnError(payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::ICD_CheckIn), CHIP_NO_ERROR);
+
+    ByteSpan payloadByteSpan{ payload->Start(), payload->DataLength() };
+    ICDClientInfo clientInfo;
+    CounterType counter = 0;
+    // If the check-in message processing fails, return CHIP_NO_ERROR and exit.
+    VerifyOrReturnError(CHIP_NO_ERROR == mpICDClientStorage->ProcessCheckInPayload(payloadByteSpan, clientInfo, counter),
+                        CHIP_NO_ERROR);
+    CounterType receivedCheckInCounterOffset = (counter - clientInfo.start_icd_counter) % kCheckInCounterMax;
+
+    // Detect duplicate check-in messages and return CHIP_NO_ERROR on receiving a duplicate message
+    if (receivedCheckInCounterOffset <= clientInfo.offset)
+    {
+        ChipLogError(ICD, "A duplicate check-in message was received and discarded");
+        return CHIP_NO_ERROR;
+    }
+
+    clientInfo.offset = receivedCheckInCounterOffset;
+    bool refreshKey   = (receivedCheckInCounterOffset > kKeyRefreshLimit);
+
+    if (refreshKey)
+    {
+        uint8_t newKeyData[chip::Crypto::kAES_CCM128_Key_Length];
```

Comment:
Yeah, so this newKeyData... it's unused.

Reply:
Yes, with the store and retrieve logic, this is unused. We don't have to pass anything to the application here. Will update this.